interface/src/app/sample_annotation/annotation.js

```diff
@@ -0,0 +1,129 @@
+        self.hasNode = true;
+        self.precision = 5;
+        var nodeID = $stateParams.node;
+        Activity.get(
+          {'node': nodeID, 'sample__in': samplesInUrl},  // parameters of GET
+          function success(response) {
+            var validSamples = [];
+            response.objects.forEach(function(element) {
+              validSamples.push(element.sample);
+              sampleDict[element.sample] = {activity: element.value};
+            });
+            deferred.resolve(validSamples.join());
+          },
+          function error(response) {
+            var errMessage = errGen('Failed to get activities', response);
+            $log.error(errMessage);
+            self.queryStatus = errMessage + '. Please try again later.';
+            deferred.reject(errMessage);
+          });
+      } else { // If 'node' is not in the URL, include all input sample IDs.
+        deferred.resolve(samplesInUrl);
+      }
+      return deferred.promise;
+    };
+
+    var sidPromise = getSampleIDPromise();
+    // The following then() block only includes a success handler function,
+    // because the error has been reported by setting self.queryStatus.
+    sidPromise.then(function success(validSamples) {
+      Sample.get(
+        {'id__in': validSamples, 'limit': 0}, // parameters of GET
+        function success(response) {
+          // If no valid sample records are found, report the error.
+          if (!response.objects.length) {
+            self.queryStatus = 'Valid samples not found';
+            return;
+          }
+          response.objects.forEach(function(element) {
+            if (!self.hasNode) { // If node is not in URL, create the object.
+              sampleDict[element.id] = {};
+            }
+            sampleDict[element.id].name = element.name;
+            sampleDict[element.id].annotations = element.annotations;
+            var currentTypes = Object.keys(element.annotations);
+            currentTypes.forEach(function(annotationType) {
+              if (self.uniqueAnnotationTypes.indexOf(annotationType) === -1) {
+                self.uniqueAnnotationTypes.push(annotationType);
+              }
+            });
+          });
+
+          // Convert sampleDict from an object literal to an array and save
+          // it as self.samples. (It has to be an array so that the order of
+          // samples can be customized on web UI with "orderBy" filter.
+          self.samples = [];
+          Object.keys(sampleDict).map(function(key) {
+            self.samples.push(sampleDict[key]);
+          });
+          self.uniqueAnnotationTypes.sort(); // Sort the annotation types
+          self.queryStatus = '';
+        }, function error(response) {
+          var errMessage = errGen('Failed to get sample annotations', response);
+          $log.error(errMessage);
+          self.queryStatus = errMessage + '. Please try again later.';
+        });
+    }); // end of sidPromise.then()
```

Comment:
Another way to clarify the flow of deferred promises is to do the opposite of what I did above and add *more* functions, making them truly reusable, and then just calling them in series. This would highlight the "do this, then that" aspects more clearly, but requires defining each step in a functional way with inputs and return values crisply defined.